### DISCUSSION TITLE
[Question] Async validation: what has to be returned?

### DISCUSSION THREAD

**Discussion by chiester** [vazco/uniforms#1023] [Q&A] [answered]
I'm looking at the documentation at https://uniforms.tools/docs/api-forms/#asynchronous-validation

It seems that `onValidate()` has to return errors in order for the errors to be rendered in the form. What is the correct way to modify the errors to return them? 

I'm making a call to my API to do my async validation and if it fails, it's not clear to me how to pass that response back to the form. I also looked at ajv's recommended approach (https://ajv.js.org/guide/async-validation.html), but couldn't get it working in my bridge.

**Comment by radekmie** [comment, marked as answer]
Hi @chiester. As with all errors in uniforms, the format of the error is specified by the bridge you're using. So, if you're using JSON schema bridge (I assume so, based on the ajv in your message), you have to return a list of [`ErrorObject`](https://ajv.js.org/api.html#error-objects)-like objects. (I say `ErrorObject`-like, as we're not using all of the properties. You can always check the [`getError`](https://github.com/vazco/uniforms/blob/8dd2d7435d4a19c695af6e06fe0d27f97eeb7c0b/packages/uniforms-bridge-json-schema/src/JSONSchemaBridge.ts#L96-L114) implementation to see the details.)

> What is the correct way to modify the errors to return them?

In `onValidate`, you are supposed to return a new error, not modify the one you received. So, for example, in JSON schema, simply add your custom errors to the `details` array. To make them visible with the `showInlineError`, make sure that the `instancePath` is correct.

> I'm making a call to my API to do my async validation and if it fails, it's not clear to me how to pass that response back to the form. I also looked at ajv's recommended approach (https://ajv.js.org/guide/async-validation.html), but couldn't get it working in my bridge.

You don't have to use `onValidate` for that - you can use the validator itself (the function that you pass to the `JSONSchemaBridge` constructor). The only thing is that you have to return the error in a proper format (and not `reject` or `throw` it).

---

I hope this answers the question. If not, please do let me know.

**Reply by chiester** [reply]
I'm returning the error, but it's not clear to me how to show it in the <ErrorField .../>

**Reply by radekmie** [reply]
To make the error visible in `ErrorField`, you have to make it understandable for the `getError` implementation of your bridge. In general, it has to have the `name` of the field. _(For details, refer to the `getError` implementation in your bridge.)_

**Comment by mariusrak** [comment]
Hi I'm also battling with this. Again. Can we please have noted the exact schema of error object somewhere in documentation? Thanks

**Reply by wadamek65** [reply]
@mariusrak I believe this is something we already touched upon and we've realized that we most likely can introduce stricter types for the errors inside the bridges directly so this should be an even better improvement. If for some reason that will not come to fruition, we will add a note in the docs.

@kestarumper ~~Can you file an issue for this as IIRC you were the one investigating this?~~ I believe the issue is this one #1158.
